src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionCommitManager.cs

```diff
@@ -0,0 +1,170 @@
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Remove:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return false;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Replace:
+                        return rule.Characters.Contains(typeChar);
+                }
+            }
+
+            return s_commitChars.Contains(typeChar);
+        }
+
+        private EditorCompletion.CommitBehavior CustomCommit(
+            Document document,
+            IFeaturesCustomCommitCompletionProvider provider,
+            ITextView view,
+            ITextBuffer subjectBuffer,
+            RoslynCompletionItem roslynItem,
+            ITrackingSpan applicableSpan,
+            char commitCharacter,
+            CancellationToken cancellationToken)
+        {
+            bool includesCommitCharacter;
+            if (!subjectBuffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
+                return EditorCompletion.CommitBehavior.None;
+            }
+
+            if (subjectBuffer.EditInProgress)
+            {
+                return EditorCompletion.CommitBehavior.None;
+            }
+
+            using (var edit = subjectBuffer.CreateEdit())
+            {
+                var currentViewSpan = applicableSpan.GetSpan(applicableSpan.TextBuffer.CurrentSnapshot);
+                var spans = view.BufferGraph.MapDownToBuffer(currentViewSpan, SpanTrackingMode.EdgeInclusive, subjectBuffer);
+
+                // There can be no spans in case of projections but must be less than 2 spans.
+                // 1 span is a regular case.
+                // 0 spans means the a span is in anoter projection.
+                // We do not expect more than 1 span, and do not support this case.
+                System.Diagnostics.Debug.Assert(spans.Count < 2);
+                if (spans.Any())
+                {
+                    var currentBufferSpan = spans[0].TranslateTo(subjectBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+                    edit.Delete(currentBufferSpan);
+                }
+
+                var change = provider.GetChangeAsync(document, roslynItem, commitCharacter, cancellationToken).WaitAndGetResult(cancellationToken);
+
+                // Note that we use here a snapshot created in GetCompletionContextAsync. Be sure it is up-to-date.
+                edit.Replace(change.TextChange.Span.ToSpan(), change.TextChange.NewText);
+                edit.Apply();
+
+                if (change.NewPosition.HasValue)
+                {
+                    view.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(subjectBuffer.CurrentSnapshot, change.NewPosition.Value));
+                }
+
+                includesCommitCharacter = change.IncludesCommitCharacter;
```

Comment:
i'm definitely concerned by how little this code reflects the code in Controller_Commit.cs.  Why is it so different?